File: scripts/audit-frontmatter.mjs
#!/usr/bin/env node

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const docsDir = path.join(__dirname, '..', 'docs');

/**
 * Extract frontmatter from MDX file
 * Returns { title, description, hasFrontmatter, isEmpty }
 */
function extractFrontmatter(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf-8');
    
    // Check for frontmatter delimiter
    if (!content.startsWith('---')) {
      return {
        title: null,
        description: null,
        hasFrontmatter: false,
        isEmpty: true,
        missing: ['title', 'description'],
      };
    }

    // Extract frontmatter block
    const endMatch = content.slice(3).match(/\n---\n/);
    if (!endMatch) {
      return {
        title: null,
        description: null,
        hasFrontmatter: false,
        isEmpty: true,
        missing: ['title', 'description'],
      };
    }

    const frontmatterText = content.slice(3, 3 + endMatch.index);
    const lines = frontmatterText.split('\n');
    
    let title = null;
    let description = null;

    for (const line of lines) {
      if (line.startsWith('title:')) {
        title = line.replace(/^title:\s*['"]?(.+?)['"]?\s*$/, '$1').trim();
        title = title.replace(/^['"]|['"]$/g, ''); // Remove quotes if present
      }
      if (line.startsWith('description:')) {
        description = line.replace(/^description:\s*['"]?(.+?)['"]?\s*$/, '$1').trim();
        description = description.replace(/^['"]|['"]$/g, ''); // Remove quotes if present
      }
    }

    const missing = [];
    if (!title || title === '') missing.push('title');
    if (!description || description === '') missing.push('description');

    return {
      title: title || null,
      description: description || null,
      hasFrontmatter: true,
      isEmpty: missing.length === 2,
      missing,
    };
  } catch (error) {
    console.error(`Error reading ${filePath}: ${error.message}`);
    return {
      title: null,
      description: null,
      hasFrontmatter: false,
      isEmpty: true,
      missing: ['title', 'description'],
    };
  }
}

/**
 * Recursively walk docs directory and collect all MDX files
 */
function walkDocs(dir, fileList = []) {
  const files = fs.readdirSync(dir);
  
  for (const file of files) {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    
    if (stat.isDirectory()) {
      walkDocs(filePath, fileList);
    } else if (file.endsWith('.mdx') || file.endsWith('.md')) {
      fileList.push(filePath);
    }
  }
  
  return fileList;
}

// Main audit
console.log('🔍 Frontmatter Audit\n');
console.log(`Scanning: ${docsDir}\n`);

const allFiles = walkDocs(docsDir);
const results = [];
let missingCount = 0;
let emptyCount = 0;

for (const filePath of allFiles) {
  const fm = extractFrontmatter(filePath);
  const relPath = path.relative(docsDir, filePath);
  
  results.push({
    file: relPath,
    fullPath: filePath,
    ...fm,
  });

  if (fm.missing.length > 0) {
    missingCount++;
    if (fm.isEmpty) emptyCount++;
  }
}

// Sort by file path
results.sort((a, b) => a.file.localeCompare(b.file));

// Report by status
console.log(`📊 Total MDX files: ${allFiles.length}`);
console.log(`✅ Complete: ${allFiles.length - missingCount}`);
console.log(`⚠️  Missing fields: ${missingCount}`);
console.log(`❌ Empty: ${emptyCount}\n`);

if (missingCount > 0) {
  console.log('📝 Files with missing/empty fields:\n');
  
  for (const result of results) {
    if (result.missing.length > 0) {
      console.log(`${result.file}`);
      console.log(`  Missing: ${result.missing.join(', ')}`);
      if (result.title) console.log(`  Title: "${result.title}"`);
      if (result.description) console.log(`  Description: "${result.description.substring(0, 60)}..."`);
      console.log();
    }
  }
}

// Export data for processing
const auditData = {
  timestamp: new Date().toISOString(),
  totalFiles: allFiles.length,
  completeFiles: allFiles.length - missingCount,
  missingCount,
  emptyCount,
  results: results.map(r => ({
    file: r.file,
    fullPath: r.fullPath,
    title: r.title,
    description: r.description,
    missing: r.missing,
  })),
};

fs.writeFileSync(
  path.join(__dirname, 'audit-frontmatter.json'),
  JSON.stringify(auditData, null, 2)
);

console.log(`💾 Audit data saved to: scripts/audit-frontmatter.json`);

// Exit with code 1 if there are missing fields (for CI/CD)
process.exit(missingCount > 0 ? 1 : 0);

File: audit-frontmatter.py
#!/usr/bin/env python3
import os
import re
import json
from pathlib import Path

docs_dir = Path('docs')

def extract_frontmatter(file_path):
    """Extract title and description from MDX frontmatter"""
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Check for frontmatter
    if not content.startswith('---'):
        return {
            'title': None,
            'description': None,
            'hasFrontmatter': False,
            'missing': ['title', 'description']
        }
    
    # Extract frontmatter block
    match = re.match(r'^---\n(.*?)\n---\n', content, re.DOTALL)
    if not match:
        return {
            'title': None,
            'description': None,
            'hasFrontmatter': False,
            'missing': ['title', 'description']
        }
    
    frontmatter_text = match.group(1)
    title = None
    description = None
    
    # Parse YAML-like frontmatter
    for line in frontmatter_text.split('\n'):
        if line.startswith('title:'):
            title = line.replace('title:', '').strip()
            # Remove quotes if present
            title = re.sub(r'^["\']|["\']$', '', title)
        elif line.startswith('description:'):
            description = line.replace('description:', '').strip()
            # Remove quotes if present
            description = re.sub(r'^["\']|["\']$', '', description)
    
    missing = []
    if not title or title == '':
        missing.append('title')
    if not description or description == '':
        missing.append('description')
    
    return {
        'title': title,
        'description': description,
        'hasFrontmatter': True,
        'missing': missing
    }

# Walk all MDX files
all_files = []
for root, dirs, files in os.walk(docs_dir):
    for file in sorted(files):
        if file.endswith('.mdx') or file.endswith('.md'):
            all_files.append(os.path.join(root, file))

# Audit each file
results = []
missing_count = 0
empty_count = 0

for file_path in all_files:
    fm = extract_frontmatter(file_path)
    rel_path = os.path.relpath(file_path, docs_dir)
    
    if fm['missing']:
        missing_count += 1
        if len(fm['missing']) == 2:
            empty_count += 1
    
    results.append({
        'file': rel_path,
        'fullPath': os.path.abspath(file_path),
        **fm
    })

# Print report
print(f"🔍 Frontmatter Audit\n")
print(f"📊 Total MDX files: {len(all_files)}")
print(f"✅ Complete: {len(all_files) - missing_count}")
print(f"⚠️  Missing fields: {missing_count}")
print(f"❌ Empty: {empty_count}\n")

if missing_count > 0:
    print("📝 Files with missing/empty fields:\n")
    for result in results:
        if result['missing']:
            print(f"{result['file']}")
            print(f"  Missing: {', '.join(result['missing'])}")
            if result['title']:
                print(f"  Title: \"{result['title'][:60]}{'...' if len(result['title']) > 60 else ''}\"")
            if result['description']:
                print(f"  Description: \"{result['description'][:60]}{'...' if len(result['description']) > 60 else ''}\"")
            print()

# Save audit data
audit_data = {
    'totalFiles': len(all_files),
    'completeFiles': len(all_files) - missing_count,
    'missingCount': missing_count,
    'emptyCount': empty_count,
    'results': results
}

with open('audit-frontmatter.json', 'w') as f:
    json.dump(audit_data, f, indent=2)

print(f"💾 Audit data saved to: audit-frontmatter.json")
